test/metabase/server/middleware/offset_paging_test.clj

```diff
@@ -1,45 +1,57 @@
 (ns metabase.server.middleware.offset-paging-test
   (:require
+   [cheshire.core :as json]
    [clojure.test :refer :all]
+   [metabase.server.handler :refer [apply-middleware]]
    [metabase.server.middleware.offset-paging :as mw.offset-paging]
-   [ring.middleware.keyword-params :refer [wrap-keyword-params]]
-   [ring.middleware.params :refer [wrap-params]]
+   [metabase.server.middleware.security :as mw.security]
    [ring.mock.request :as ring.mock]
-   [ring.util.response :as response]
-   [schema.core :as s]))
+   [ring.util.response :as response])
+  (:import
+   (java.io BufferedReader InputStreamReader PipedInputStream)
+   (java.nio.charset StandardCharsets)))
 
 (defn- handler [request]
-  (let [handler* (-> (fn [request respond _]
-                       (respond (response/response {:limit  mw.offset-paging/*limit*
-                                                    :offset mw.offset-paging/*offset*
-                                                    :paged? mw.offset-paging/*paged?*
-                                                    :params (:params request)})))
-                     mw.offset-paging/handle-paging
-                     wrap-keyword-params
-                     wrap-params)
-        respond identity
-        raise   (fn [e] (throw e))]
+  (let [handler  (fn [request respond _]
+                   (respond (response/response {:limit  mw.offset-paging/*limit*
+                                                :offset mw.offset-paging/*offset*
+                                                :paged? mw.offset-paging/*paged?*
+                                                :params (:params request)})))
+        handler* (apply-middleware handler)
+        respond  identity
+        raise    (fn [e] (throw e))]
     (handler* request respond raise)))
 
+(defn- read-response
+  "Responses from our hanlders are inputstream, this is read the stream into real body."
+  [response]
+  (update response :body
+          (fn [body]
+            (if (instance? PipedInputStream body)
+              (with-open [input-reader (InputStreamReader. body StandardCharsets/UTF_8)
+                          buffer-reader (BufferedReader. input-reader)]
+                (json/parse-stream buffer-reader))
+              body))))
+
 (deftest paging-test
   (testing "no paging params"
-    (is (= {:status  200
-            :headers {}
-            :body    {:limit  nil
-                      :offset nil
-                      :paged? false
-                      :params {}}}
-           (handler (ring.mock/request :get "/")))))
+    (is (=? {:status  200
+             :headers (mw.security/security-headers)
```

Comment:
I thought about removing it but then I think it was nice to have a piece of test.

Since it's a rare case for us to have API that throws exceptions in the middleware, I'm not sure if are there any tests to make sure we have security headers in those cases.

Reply:
If so, I would put that in a separate test since it's testing a code path unrelated to offsets and pages. `=?` allows us to write narrow tests easily, so we might as well use it.

Reply:
on second thought, we explicitly include security headers when there is an error [here](https://github.com/metabase/metabase/blob/6f64abedd82d037029b8ad13ab6f214c496a475b/src/metabase/server/middleware/offset_paging.clj#L51).

So I added a headers test but only for the error case.